### DISCUSSION TITLE
We really should not care about legacy players

### DISCUSSION THREAD

**Discussion by spessasus** [SFe-Team-was-taken/SFE#34] [Ideas]
### Introduction 
One thing I noticed when suggesting features was that some of them cannot be added because they _break compatibility._

Isn't that the point of SFe? To advance and remove the limits of SF2? The SF3 format isn't backwards compatible and yet it's widely accepted.

So I think SFe should be unconstrained by the legacy sf players and the  shouldn't be taken into account when adding features. Here's why:

### What's the point?
First of all, what's the point of supporting legacy players?
It greatly limits what SFe can add. Not to mention that anything that SFe adds won't work with them.

For example I made an XG compatible sound bank using the LSB feature and added XG volume curve using default modulators.

This won't work in synth unless it's explicitly supported. Not to mention that LSB would probably make most legacy players reject the file (it would break the current Spessasynth implementation for example).

It is like playing the SF3 file with legacy players. It might open if they don't check sample type or ifil, but it will play the _compressed streams_ as raw audio, which just will be white noise. It is, of course, way more extreme than the issues mentioned above, but you get my point.

So what's the point of supporting them? At this point, you might as well feed them a regular SF2 file for the same effect.

Not to mention that legacy sf editors like Vienna for example would probably not copy the new values when saving the file. So by using `.sf2` and a backwards compatible format, we risk it getting back ported to SF2 with a legacy editor, losing all new data.

### Breaking the compatibility is actually good
It might sound strange, but it can actually benefit us. If the user sees that this sound bank does not load on their old synth or something, they will look it up and discover SFe, which will make them switch over to a supported synth, or as a developer, convince them to support SFe. 

Hell, we should even change the main header from `sfbk` to `sfeh` or something like that to intentionally make players that don't support SFes features reject the file to prevent incorrect playback.

With the format stating compatible, we make the user think that this soundfont sounds off, as the legacy player doesn't care about _anything_ SFe adds. So, again, just use an SF2 file if you want to play it on your sound blaster or something.

### too many formats!
Currently, SFe has a bunch of versions like SFe32, SFe64, SFe64L to name a few.
It's really confusing IMO. As a developer, this is what's overwhelming. I see a bunch of different versions that I need to support. Wasn't SFes point to _unify_ all soundfont formats into one? Because what currently is happening is the exact opposite.

And why do these formats exist? That's right, _backwards compatibility!_

If we didn't care about backwards compatibility, we could unify them into one format (by using my dynamic riff proposal for example). Much easier for the devs to understand, even if it's more complex than the different formats, as there would be just one SFe, and that's it!

Sorry for this wall of text, but this has been bothering me ever since I joined SFe.

**Comment by sylvia-leaf** [comment]
I've already written about plans to incorporate changes that will mean that SFe banks can't be loaded by legacy SF players [here](https://github.com/SFe-Team-was-taken/SFe/issues/35#issuecomment-2485702325), but I'm going to write some ideas on this discussion page as well. As always, if you have any feedback about these ideas, or would make any changes, please tell me; I'm all ears!

---

### How do major versions work in SFe?

The SFe major version mostly describes the **format structure** of the format. SFe 4 will be the **only** version that will be "somewhat backwards compatible" with legacy SF2.0x (this means that it will be *loadable* on a legacy SF player that is 100% legacy SF2.01/SF2.04 compliant). Starting from SFe 5, the file format will be incompatible with previous major versions, but each major version will remain superficially similar to the previous version. We don't want to create two almost completely different file formats with a name that only differs in terms of major version - SFe 4 and SFe 5 will be relatively similar to each other.

Changes in the file format that are incompatible will be made gradually so it is easy for developers to adapt and create programs that support future major versions of SFe. It also allows files to be converted between versions more easily, by taking steps and using other major versions as intermediates between the input and output. While it should eventually be possible to make a program that converts to and from different versions of SFe in one pass, it might be easier for a program to convert between versions by upgrading or downgrading one version at a time.

### Long term support (LTS) and SFe

Another advantage of this approach that I've mentioned before is that we can define the most popular SFe major versions as "long term support" (LTS), a term often used in open-source software. But we can go beyond a typical LTS system; we can add features that we come up with retroactively to previous versions, even after a new major version is released. So, after releasing SFe 5, you can expect that more features will be added to SFe 4. The only thing is that features that depend on structural changes that distinguish two major versions won't be included. So, while SFe 4 may eventually get SFe 5 features in the future, what it *won't* get (for example) is unlimited sample/instrument/preset name lengths, as the 20-character limit is hardcoded into the legacy SF format.

This was the initial idea behind SFe32 and SFe64. SFe32 would remain "somewhat backwards compatible" for a program that is completely compliant with legacy SF, and SFe64 would have the incompatible changes. However, the idea of having two different specifications with such similar names being maintained simultaneously was determined to be too complicated, and was also against the spirit of SFe. By combining SFe32 and SFe64 into one specification using the same structure but different chunk headers, we simplify the process of using 32-bit and 64-bit headers on SFe. We can also add in the future even more chunk header types, such as the "dynamic RIFF" chunk header concept.

This also solves the potential problem of SFe software that only supports (up to) a certain major version (for example SFe 4) becoming very popular, resulting in possible new features being withheld. By independently updating each version when possible, we can maximise the number of features available in a given major version to reduce the requirement for SFe software developers to rewrite their software every so often. We can also release new major versions faster and add essential features as they become required, while giving those who are on an older SFe major version time to migrate to the newest version. It may even be feasible for a developer to skip major versions and go straight to a new version (for example, skipping SFe 5 and updating a program from SFe 4 directly to SFe 6).

And while we can add features to older versions of SFe, we will always prioritise the latest version, because ideally, we want everyone who implements SFe to be on the latest version. The LTS system is really intended for the developers that are stubborn and don't want to switch to a newer version due to an impending rewrite (for example). Eventually, we may end support for obsolete SFe versions.

### The problem with legacy editors

If you use a legacy editor, then you are right - SFe-specific data may just be lost. This means that if you attempt to playback the edited SFe bank on an SFe player, all of the new features will no longer be usable.

The mitigation for this issue is to make sure that any SFe bank that contains extra data that may be removed when edited with a legacy editor uses the file extension `.sft`, as specified in the SFe specification. To even open the SFe bank on a legacy editor, then the user will have to change the file extension from `.sft` to `.sf2`. Do you think that this is a sufficient failsafe?

### Main header modifications

For SFe 5 and later (not "somewhat backwards compatible" due to significant structural changes), and SFe 4 with a chunk header that is not "static 32-bit", the data may be read incorrectly by a legacy SF player. Because of this, there may be sample, instrument or preset corruption if an attempt is made to load such a SFe bank in a legacy player.

The mitigation as suggested by you is to rename the `sfbk` header to something else. So, we've decided to use `sfbk` only for SFe 4 with static 32-bit chunk headers, and `sfen` for all other SFe banks.

### `sfbk` stubs and what they can do with the format

In `SFSPEC24.PDF`, section 4.1 shows that the three defined chunks are included in the `sfbk` header, but it may be possible to have another header beyond `sfbk`. Hypothetically, we can include both an `sfbk` and an `sfen` header. 

The idea is that the `sfbk` header includes placeholder data. For example, preset `0:0` could have the name "Unsupported Player!", preset `0:1` could have the name "SFe Player Required!", and other information can be included using the `INFO-list` chunk, as there will be two `INFO-list` chunks; the actual information can be included in the `ICMT` subchunk in `sfen`, while a link to an SFe tutorial or player can be included in the `ICMT` subchunk in `sfbk`.

Essentially, the idea is that a legacy player that is "somewhat backwards compatible" with legacy SF2.0x will read just the legacy `sfbk` formatted data and not the enhanced `sfen` formatted data, and then give a customised error to the user. It won't affect the use of the bank for players that aren't 100% compliant with SF2.04, but it will help users to get the bank working. The main risk is that a badly created program may attempt to load the whole file, see two `RIFF` headers and error out. If this happens, then there isn't a change between a file that has a `sfbk` stub and a file that does not.

Should there be a legacy SF player that doesn't error out on a second `RIFF` header, then not only does the stub work properly and can give instructions on how to get an SFe player that works properly with the bank, but bank developers could potentially build different versions of the bank that are loaded depending on whether it is loaded in legacy SF or SFe. The only problem is that samples will be duplicated, and that is not something that can easily be fixed. Therefore, having both `sfbk` and `sfen` is intended to allow the bank developer to include a stub to give information on how to load SFe. Think of it as the SFe version of the "DOS stub" (the `This program can't be run in DOS mode` message that you see on almost all Windows executables).

### Yet another fourcc for RIFF 

So, standards organisations have defined `RIFF` and `RF64` already, and the "dynamic RIFF" concept will use the header `RIFD`. Big-endian versions of RIFF use the header `RIFX`.

For the person who uses SFe with a big-endian operating system, we'll officially allow users to use `RIFX` as their header. Support for big-endian is completely optional, and won't even be in-specification until version 4.1 or later. The big-endian feature flag must also be recognised by the player for it to work properly.
